### DISCUSSION TITLE
Unable to retrieve token {:error, {:bad_body_generator, {:error, :invalid_input, :none}}}

### DISCUSSION THREAD

**Discussion by sigu** [erlef/oidcc_plug#28] [Q&A] [answered]
oidcc_plug version

0.1.2
Erlang version

27.1
Elixir version

1.17.3-otp-27

I am getting the error above from a keycloak local instance.
After inspecting through on [these lines](https://github.com/erlef/oidcc_plug/blob/c195e9d4d3e11eebf115b97bcea3142fb6652085/lib/oidcc/plug/authorization_callback.ex#L200), here are the parameters that gets calculated right before requesting for the token:

```
CODE: "1da4d921-6772-4c2d-b1e6-d23ee593c4ca.84dee313-0256-4663-a434-4747949c52a0.29105a68-a2c5-4a7c-ac6b-770a76eca034"
SCOPE: "openid"
OPTS: %{
  scope: ["openid"],
  redirect_uri: "http://localhost:4010/oidcc/callback",
  request_opts: %{},
  pkce_verifier: :none,
  nonce: :any
}
```

Here is the error

`{:error, {:bad_body_generator, {:error, :invalid_input, :none}}}`

I suspect the keycloak server is refusing the `:none` value and I am not sure which values I can successfully pass

### More configs that I have
1. starting up the worker
```
        {Oidcc.ProviderConfiguration.Worker,
         %{
           issuer: "http://localhost:7080/realms/master",
           name: MyApp.KeycloakConfigurationProvider,
           provider_configuration_opts: %{
             quirks: %{
               allow_unsafe_http: true,
               document_overrides: %{
                 "token_endpoint_auth_methods_supported" => ["client_secret_basic"],
                 "introspection_endpoint_auth_methods_supported" => ["client_secret_basic"]
               }
             }
           }
```
2. keycloak config
The keycloak client is a confidential client that uses the secret key

**Comment by maennchen** [comment]
@sigu I believe the error does not come from Keycloak. Instead it happens even before that.

I see that your options include `pkce_verifier: :none`. That is not a valid option since it should either be a string or not be present at all. Can you try just leaving it out?

**Reply by sigu** [reply]
I get an error too `PKCE code verifier not specified` 


```
CODE: "c3ba47e0-2ea4-45dc-b945-2e73598c77ce.4256f52a-bac1-49e6-b763-c9a77365e164.29105a68-a2c5-4a7c-ac6b-770a76eca034"
SCOPE: "openid"
OPTS: %{
  scope: ["openid"],
  redirect_uri: "http://localhost:4010/oidcc/callback",
  request_opts: %{},
  nonce: :any
}
{:error,
 {:http_error, 400,
  %{
    "error" => "invalid_grant",
    "error_description" => "PKCE code verifier not specified"
  }}}
```

**Reply by sigu** [reply]
the `pkce_verifier` option is also defined on the library [open link](https://github.com/erlef/oidcc_plug/blob/c195e9d4d3e11eebf115b97bcea3142fb6652085/lib/oidcc/plug/authorization_callback.ex#L213)

**Reply by maennchen** [reply]
Ok, that means that Keycloak expects you to add one. `:none` is not a valid option for that.

Essentially, you will have to generate a PKCE Verifier when you create the authorization request and specify it again when you exchange the token.

`oidcc_plug` does this as well. Have a look at the implementation there:

https://github.com/erlef/oidcc_plug/blob/c195e9d4d3e11eebf115b97bcea3142fb6652085/lib/oidcc/plug/authorize.ex
https://github.com/erlef/oidcc_plug/blob/c195e9d4d3e11eebf115b97bcea3142fb6652085/lib/oidcc/plug/authorization_callback.ex

**Reply by sigu** [reply]
I am using `oidcc_plug` on my code base, the inspects are actually inside the `oidcc_plug` deps directory. How can I enable the application to generate the PKCE?

**Reply by maennchen** [reply]
> the `pkce_verifier` option is also defined on the library [open link](https://github.com/erlef/oidcc_plug/blob/c195e9d4d3e11eebf115b97bcea3142fb6652085/lib/oidcc/plug/authorization_callback.ex#L213)

Oh, that's probably a bug. Are you calling the oidcc functions yourself or are you debugging the plugs?

**Reply by sigu** [reply]
I am debugging the plugs

**Reply by maennchen** [reply, marked as answer]
Can you open an issue (or a PR if you're up to the task) in the other repo? I'll have a look.

Essentially we'll just have to remove `none` values when calling the function.

**Reply by sigu** [reply]
I have converted this into an issue, will open a PR later

**Reply by GPrimola** [reply]
Hi @sigu, I'm very interested in this fix, let me know if you need any help to speed things up :)

Thank you very much!

**Reply by maennchen** [reply]
Issue here: https://github.com/erlef/oidcc_plug/issues/29

@GPrimola If you want to give it a go, that also works for me. It would be good to get a fix out but I'm a bit swamped myself ATM.

**Reply by maennchen** [reply]
Fixed and release as v0.2.0

**Reply by sigu** [reply]
Thank you for fixing this @maennchen
